### ISSUE TITLE
Better error message for malformed item transformations

### ISSUE THREAD

**Issue by cdmihai** [Microsoft/msbuild#333]
A definition of an item `ThisItem`

`<ItemGroup>
    <ThisItem Include="@(AnotherItem->$(APath)\%(Filename)%(Extension))" />
</ItemGroup>`

Followed by a metadata access of `ThisItem` of the form:

`%(ThisItem.FullPath)`

Failes with the error: *error MSB4023: Cannot evaluate the item metadata "%(FullPath)". The item metadata "%(FullPath)" cannot be applied to the path "@(AnotherItem->c:\projects\msbuild\bin\Debug-NetCore\Windows_NT_Deployment_Test\)". Illegal characters in path.*

The error is currently caught when the malformed item is used, not when the malformed item is created. Instead, the error should be caught at item definition with a reference to the fact that there are no single quotes (correct definition is "@(AnotherItem->'$(APath)\%(Filename)%(Extension)')")

**Comment by rainersigwald** [comment]
Note that the reason this isn't an error now is that MSBuild's model is permissive.  When it determines that the item transformation isn't valid (it's either missing quotes--`@(AnotherItem->'$(APath)\%(Filename)%(Extension)')`--or it's missing an item function name after the `->`), it falls back to interpreting what it can as literal text (`@(AnotherItem->`), property expansions (`$(APath)` → `c:\projects\msbuild\bin\Debug-NetCore\Windows_NT_Deployment_Test`), and shared metadata (`%(Filename)` and `%(Extension)` without an item qualifier, which both expand to the empty string).

There are even tests to enforce this behavior, such as [`ExpandItemVectorFunctionsInvalid2`](https://github.com/Microsoft/msbuild/blob/a4e0dd5f4b31a6c9acb1bab25ac401c438c3dfac/src/XMakeBuildEngine/UnitTests/Evaluation/Expander_Tests.cs#L295)!

Because this permissiveness is a fairly core property of MSBuild, we can't take the approach I would really want (a parse error when considering the item in the first place).  We also probably can't introduce a hard error for this situation, because it would be easy for many customers to have taken an inadvertent dependency on the current behavior.  I think the best we could do is emit a warning.

**Comment by cdmihai** [comment]
I agree. In this case, for the sake of internal consistency, it would be better to have a warning when there's a missing item qualifier.
